sparrow-py/src/execution.rs

```diff
@@ -1,44 +1,97 @@
-pub(crate) struct Execution(Option<RustExecution>);
+#[derive(Clone)]
+pub(crate) struct Execution {
+    execution: Arc<Mutex<Option<RustExecution>>>,
+}
 
 impl Execution {
     pub(crate) fn new(execution: RustExecution) -> Self {
-        Self(Some(execution))
+        Self {
+            execution: Arc::new(Mutex::new(Some(execution))),
+        }
+    }
+}
+
+impl Execution {
+    /// Return a mutex-locked `RustExecution`.
+    ///
+    /// Error if the execution has already been completed (consumed).
+    fn execution(&self) -> Result<MutexGuard<'_, Option<RustExecution>>> {
+        let execution_opt = self.execution.blocking_lock();
+        if execution_opt.is_none() {
+            return Err(error_stack::report!(ErrorContext::ResultAlreadyCollected).into());
+        }
+        Ok(execution_opt)
+    }
+
+    /// Same as `execution` but returning an `OwnedMutexGuard` which is `Send`.
+    async fn owned_execution(&self) -> Result<OwnedMutexGuard<Option<RustExecution>>> {
+        let execution_opt = self.execution.clone().lock_owned().await;
+        if execution_opt.is_none() {
+            return Err(error_stack::report!(ErrorContext::ResultAlreadyCollected).into());
+        }
+        Ok(execution_opt)
+    }
+
+    fn take_execution(&self) -> Result<RustExecution> {
+        Ok(self.execution()?.take().unwrap())
     }
 }
 
+impl Execution {}
+
 #[pymethods]
 impl Execution {
-    fn collect_pyarrow(&mut self, py: Python<'_>) -> Result<Vec<PyObject>, Error> {
-        let execution = self
-            .0
-            .take()
-            .ok_or_else(|| error_stack::report!(ErrorContext::ResultAlreadyCollected))?;
+    fn collect_pyarrow(&mut self, py: Python<'_>) -> Result<Vec<PyObject>> {
+        let execution = self.take_execution()?;
         let batches = execution.collect_all_blocking()?;
         let results = batches
             .into_iter()
             .map(|batch| batch.to_pyarrow(py))
-            .collect::<Result<Vec<_>, _>>()?;
+            .collect::<PyResult<Vec<_>>>()?;
         Ok(results)
     }
 
-    fn next_pyarrow(&mut self, py: Python<'_>) -> Result<Option<PyObject>, Error> {
-        let execution = self
-            .0
-            .as_mut()
-            .ok_or_else(|| error_stack::report!(ErrorContext::ResultAlreadyCollected))?;
-        let batch = execution.next_blocking()?;
+    fn next_pyarrow(&mut self, py: Python<'_>) -> Result<Option<PyObject>> {
+        let mut execution = self.execution()?;
+        let batch = execution.as_mut().unwrap().next_blocking()?;
         let result = match batch {
             Some(batch) => Some(batch.to_pyarrow(py)?),
             None => None,
         };
         Ok(result)
     }
+
+    fn next_pyarrow_async<'py>(&self, py: Python<'py>) -> PyResult<&'py PyAny> {
+        let execution = self.clone();
+        pyo3_asyncio::tokio::future_into_py(py, async move {
+            // We can't use `?` here because it attempts ot acquire the GIL unexpectedly.
```

Comment:
We can just wait a few years and they'll have removed the GIL! https://discuss.python.org/t/a-steering-council-notice-about-pep-703-making-the-global-interpreter-lock-optional-in-cpython/30474

Reply:
Maybe? not sure if that will always be removed -- and it will change a good deal of things (eg., right now we need to get the lock to allocate, but then we know that we are the only one interacting with Python stuff).